ledger/ledger-on-memory/src/main/scala/com/daml/ledger/on/memory/InMemoryLedgerReaderWriter.scala

```diff
@@ -53,39 +49,58 @@ final class InMemoryLedgerReaderWriter(
-      currentState.synchronized {
-        val stateInputs: Map[DamlStateKey, Option[DamlStateValue]] =
-          submission.getInputDamlStateList.asScala
-            .map(key => key -> currentState.state.get(key.toByteString))(breakOut)
-        val entryId = sequentialLogEntryId.next()
-        val (logEntry, damlStateUpdates) =
-          KeyValueCommitting.processSubmission(
-            engine,
-            entryId,
-            currentRecordTime(),
-            LedgerReader.DefaultConfiguration,
-            submission,
-            participantId,
-            stateInputs,
-          )
-        val stateUpdates = damlStateUpdates.map {
-          case (damlStateKey, value) => damlStateKey.toByteString -> value
+  private class InMemoryLedgerStateAccess(theParticipantId: ParticipantId)
+      extends LedgerStateAccess {
+    override def inTransaction[T](body: LedgerStateOperations => Future[T]): Future[T] =
+      Future {
+        currentState.state.synchronized {
+          body(new InMemoryLedgerStateOperations)
         }
-        currentState.log += LogEntry(entryId, Envelope.enclose(logEntry).toByteArray)
-        currentState.state ++= stateUpdates
-        dispatcher.signalNewHead(currentState.log.size)
+      }.flatten
+
+    override def participantId: String = theParticipantId
+  }
+
+  private class InMemoryLedgerStateOperations extends BatchingLedgerStateOperations {
+    override def readState(keys: Seq[Array[Byte]]): Future[Seq[Option[Array[Byte]]]] =
+      Future.successful {
+        keys.map(keyBytes => currentState.state.get(ByteString.copyFrom(keyBytes)))
       }
-      SubmissionResult.Acknowledged
-    }
+
+    override def writeState(keyValuePairs: Seq[(Array[Byte], Array[Byte])]): Future[Unit] =
+      Future.successful {
+        currentState.state ++= keyValuePairs.map {
+          case (keyBytes, valueBytes) => ByteString.copyFrom(keyBytes) -> valueBytes
+        }
+      }
+
+    override def appendToLog(key: Array[Byte], value: Array[Byte]): Future[Unit] =
+      Future.successful {
+        val damlLogEntryId = KeyValueCommitting.unpackDamlLogEntryId(key)
+        val logEntry = LogEntry(damlLogEntryId, value)
+        val newHead = currentState.log.synchronized {
+          currentState.log += logEntry
+          currentState.log.size
+        }
+        dispatcher.signalNewHead(newHead)
+      }
+  }
+
+  private val validator = SubmissionValidator.create(
+    new InMemoryLedgerStateAccess(participantId),
+    () => sequentialLogEntryId.next())
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] = {
+    validator
+      .validateAndCommit(envelope, correlationId, currentRecordTime())
+      .map {
+        case SubmissionValidated => SubmissionResult.Acknowledged
+        case MissingInputState(_) => SubmissionResult.InternalError("Missing input state")
+        case ValidationError(reason) => SubmissionResult.InternalError(reason)
+      }
+  }
```

Comment:
This seems like boilerplate; can it be moved to KVUtils?

Reply:
Good point -- let's move it when we move another ledger to use `SubmissionValidator`.